client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -250,33 +250,71 @@ where
 	}
 }
 
+/// Internal representation of the initial blocks that should be
+/// reported or ignored by the chainHead.
+#[derive(Clone, Debug)]
+struct InitialBlocks<Block: BlockT> {
+	/// Children of the latest finalized block, for which the `NewBlock`
+	/// event must be generated.
+	///
+	/// It is a tuple of (block hash, parent hash).
+	in_memory: Vec<(Block::Hash, Block::Hash)>,
+	/// Blocks that should not be reported as pruned by the `Finalized` event.
+	///
+	/// The substrate database will perform the pruning of height N at
+	/// the finalization N + 1. We could have the following block tree
+	/// when the user subscribes to the `unfollow` method:
+	///   [A] - [A1] - [A2] - [A3]
+	///                 ^^ finalized
+	///       - [A1] - [B1]
+	///
+	/// When the A3 block is finalized, B1 is reported as pruned, however
+	/// B1 was never reported as `NewBlock` (and as such was never pinned).
+	/// This is because the `NewBlock` events are generated for children of
+	/// the finalized hash.
+	pruned_forks: HashSet<Block::Hash>,
+}
+
 /// Get the in-memory blocks of the client, starting from the provided finalized hash.
-///
-/// Returns a tuple of block hash with parent hash.
-fn get_initial_blocks<BE, Block>(
+fn get_initial_blocks_with_forks<Block, BE>(
 	backend: &Arc<BE>,
-	parent_hash: Block::Hash,
-) -> Vec<(Block::Hash, Block::Hash)>
+	finalized: HashAndNumber<Block>,
+) -> Result<InitialBlocks<Block>, SubscriptionManagementError>
 where
 	Block: BlockT + 'static,
 	BE: Backend<Block> + 'static,
 {
-	let mut result = Vec::new();
-	let mut next_hash = Vec::new();
-	next_hash.push(parent_hash);
-
-	while let Some(parent_hash) = next_hash.pop() {
-		let Ok(blocks) = backend.blockchain().children(parent_hash) else {
-			continue
-		};
-
-		for child_hash in blocks {
-			result.push((child_hash, parent_hash));
-			next_hash.push(child_hash);
+	let blockchain = backend.blockchain();
+	let leaves = blockchain
+		.leaves()
+		.map_err(|err| SubscriptionManagementError::Custom(err.to_string()))?;
+	let finalized_number = finalized.number;
+	let finalized = finalized.hash;
+	let mut pruned_forks = HashSet::new();
+	let mut in_memory = Vec::new();
```

Comment:
The issue would be that `HashSet` stores elements in arbitrary order, taken from [here](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.iter). 

> An iterator visiting all elements in arbitrary order

We'd need to return the blocks in order, because the spec states that parent of `NewBlock` was prior reported by (`NewBlock` or by the `Initial` events).

Reply:
The elements in a BTreeSet will still be sorted in ascending order, which may not necessarily match the order in which they were inserted. However, I was able to remove the dependency on itertools by using an auxiliary hash set. Thanks!